vispy/visuals/filters/clipping_planes.py

```diff
@@ -0,0 +1,95 @@
+from .base_filter import Filter
+
+
+class PlanesClipper(Filter):
+    """Clips visual output based on arbitrary clipping planes."""
+
+    VERT_CODE = """
+    void clip() {
+        // Transform back to visual coordinates and clip based on that
+        $v_is_shown = $clip_with_planes($itransform(gl_Position).xyz);
+    }
+    """
+
+    FRAG_CODE = """
+    void clip() {
+        if ($v_is_shown < 0.)
+            discard;
+    }
+    """
+
+    def __init__(self, clipping_planes=None):
+        super().__init__(
+            vcode=Function(self.VERT_CODE), vhook='post', vpos=1,
+            fcode=Function(self.FRAG_CODE), fhook='pre', fpos=1,
+        )
+
+        v_is_shown = Varying('v_is_shown', 'float')
+        self.vshader['v_is_shown'] = v_is_shown
+        self.fshader['v_is_shown'] = v_is_shown
+
+        self.clipping_planes = clipping_planes
+
+    @staticmethod
+    def _get_itransform(visual):
+        """
+        get the transform from gl_Position to visual coordinates
+        """
+        return visual.get_transform('render', 'visual')
+
+    def _attach(self, visual):
+        super()._attach(visual)
+        self.vshader['itransform'] = self._get_itransform(visual)
+
+    @staticmethod
+    @lru_cache(maxsize=10)
+    def _build_clipping_planes_func(n_planes):
+        """Build the code snippet used to clip the volume based on self.clipping_planes."""
+        func_template = '''
+            float clip_planes(vec3 loc) {{
+                float is_shown = 3.4e38; // max float
+                {clips};
+                return is_shown;
+            }}
+        '''
+        # the vertex is considered clipped if on the "negative" side of the plane
+        clip_template = '''
+            vec3 relative_vec{idx} = loc - $clipping_plane_pos{idx};
+            float is_shown{idx} = dot(relative_vec{idx}, $clipping_plane_norm{idx});
+            is_shown = min(is_shown{idx}, is_shown);
+            '''
+        all_clips = []
+        for idx in range(n_planes):
+            all_clips.append(clip_template.format(idx=idx))
+        formatted_code = func_template.format(clips=''.join(all_clips))
+        return Function(formatted_code)
+
+    @property
+    def clipping_planes(self):
+        """Get the set of planes used to clip the mesh.
+        Each plane is defined by a position and a normal vector (magnitude is irrelevant). Shape: (n_planes, 2, 3)
+        """
+        return self._clipping_planes[:, :, ::-1]
+
+    @clipping_planes.setter
+    def clipping_planes(self, value):
+        if value is None:
+            value = np.empty([0, 2, 3])
+        value = value[:, :, ::-1]
+        self._clipping_planes = value
+
+        clip_func = self._build_clipping_planes_func(len(value))
+        self.vshader['clip_with_planes'] = clip_func
+
+        for idx, plane in enumerate(value):
+            clip_func[f'clipping_plane_pos{idx}'] = tuple(plane[0])
+            clip_func[f'clipping_plane_norm{idx}'] = tuple(plane[1])
```

Comment:
I can't remember how this works again. Does this mean that the program will be re-compiled each time that the clip plane is changed?

Reply:
no, recompilation only happens if the number of clipping planes is changed. This is an ugly workaround cause I could not find a way to pass a variable length 2D array to glsl. AFAIK, the only current way would be to use a large buffer and use the first value to say how many planes to consider... but it seemed overkill.

Reply:
@brisvag I agree with you that this particular set of lines that changes the `pos` and `norm` should cause a recompilation, but the lines above where you recreate the `clip_func` function and then set it again, I think **that** causes the shader to recompile. May need to do something like `if len(value) != len(self._clipping_planes): # recreate func`

Reply:
Right... I thought that the use of caching in `_build_clipping_planes_func` would be enough to prevent wasted computation. Does dollar-sign-variable assignment trigger recompilation even when the new value is the same as the previous value?

Reply:
Oh! I forgot about the lru_cache. I'm not sure if it checks if the value is the same.

Reply:
I *think* these lines do check for that:

https://github.com/vispy/vispy/blob/5d58b832c853c45b44f5036b9e7c90d6e9f4f454/vispy/visuals/shaders/function.py#L221-L223

`Function` has no `__eq__` method, but since the cache actually returns the same object, this should work fine

Reply:
Looks good to me :+1: Nice detective work.